src/java.base/share/classes/java/util/zip/GZIPInputStream.java

```diff
@@ -76,10 +88,75 @@ private void ensureOpen() throws IOException {
+
+    /**
+     * Creates a new input stream with the specified buffer size,
+     * support for concatenated streams, and tolerance for extra bytes.
+     *
+     * <p>
+     * When {@code allowConcatenation} is true, this class will attempt to decode any data that
+     * follows a GZIP trailer frame as the GZIP header frame of a new compressed data stream
+     * and proceed to decompress it. As a result, arbitrarily many consecutive compressed data
+     * streams in the underlying input will be read back as a single uncompressed stream. When
+     * {@code allowConcatenation} is false, decompression stops after the first GZIP trailer
+     * frame encountered, and any additional bytes are considered extraneous.
+     *
+     * <p>
+     * The {@code ignoreExtraBytes} flag controls the behavior when extraneous data appears,
+     * i.e., when any data follows the GZIP trailer frame (if {@code allowConcatenation} is
+     * false), or when any data other than a valid GZIP header frame follows a GZIP trailer
+     * frame (if {@code allowConcatenation} is true). When {@code ignoreExtraBytes} is true,
+     * the unexpected data, and any {@link IOException} thrown while trying to read that
+     * data, are simply discarded, and EOF is returned; when {@code ignoreExtraBytes} is
+     * false, any unexpected data triggers an {@link IOException}, and any {@link IOException}
+     * thrown while trying to read a concatenated GZIP header frame is propagated to the caller.
+     * In other words, every byte of the underlying data stream is part of a complete and valid
+     * compressed data stream, or else an {@link IOException} is guaranteed to be thrown.
+     *
+     * @apiNote The original behavior of this class is replicated by setting both
+     * {@code allowConcatenation} and {@code ignoreExtraBytes} to true. However,
+     * enabling {@code ignoreExtraBytes} is discouraged because of its imprecision in
+     * how many additional bytes are read and the possibility that {@link IOException}s
+     * and data corruption in the underlying stream can go unnoticed.
+     *
+     * @param in the input stream
+     * @param size the input buffer size
+     * @param allowConcatenation true to allow multiple concatenated compressed data streams,
+     *                           or false to expect exactly one compressed data stream
+     * @param ignoreExtraBytes true to tolerate and ignore extra bytes, false to throw
+     *                             {@link IOException} if any extra bytes are encountered
+     *
+     * @throws    ZipException if a GZIP format error has occurred or the
+     *                         compression method used is unsupported
+     * @throws    NullPointerException if {@code in} is null
+     * @throws    IOException if an I/O error has occurred
+     * @since     24
+     */
+    public GZIPInputStream(InputStream in, int size,
+            boolean allowConcatenation, boolean ignoreExtraBytes) throws IOException {
```

Comment:
I haven't reviewed the javadoc changes. I will do that separately once we settle down on the API and implementation changes.
As for this new constructor, I think the only new parameter we should introduce is whether or not the underlying input stream `in` is expected/allowed to have concatenated GZIP stream(s). So I think we should remove the "ignoreExtraBytes" new parameters from this constructor and. Additionally, I think the `allowConcatenation` should instead be named `allowConcatenatedGZIPStream`:

```java
public GZIPInputStream(InputStream in, int size, boolean allowConcatenatedGZIPStream) throws IOException {
```

Reply:
Just to provide a more concrete input, here's a very minimal tested version of what I had in mind:

```diff
--- a/src/java.base/share/classes/java/util/zip/GZIPInputStream.java
+++ b/src/java.base/share/classes/java/util/zip/GZIPInputStream.java
@@ -55,6 +55,8 @@ public class GZIPInputStream extends InflaterInputStream {
 
     private boolean closed = false;
 
+    private final boolean allowConcatenatedGZIPStream;
+
     /**
      * Check to make sure that this stream has not been closed
      */
@@ -76,14 +78,7 @@ private void ensureOpen() throws IOException {
      * @throws    IllegalArgumentException if {@code size <= 0}
      */
     public GZIPInputStream(InputStream in, int size) throws IOException {
-        super(in, createInflater(in, size), size);
-        usesDefaultInflater = true;
-        try {
-            readHeader(in);
-        } catch (IOException ioe) {
-            this.inf.end();
-            throw ioe;
-        }
+        this(in, size, true);
     }
 
     /*
@@ -111,7 +106,28 @@ private static Inflater createInflater(InputStream in, int size) {
      * @throws    IOException if an I/O error has occurred
      */
     public GZIPInputStream(InputStream in) throws IOException {
-        this(in, 512);
+        this(in, 512, true);
+    }
+
+    /**
+     * WIP
+     * @param in the input stream
+     * @param size the input buffer size
+     * @param allowConcatenatedGZIPStream true if the input stream is allowed to contain
+     *                                    concatenated GZIP streams. false otherwise
+     * @throws IOException if an I/O error has occurred
+     */
+    public GZIPInputStream(InputStream in, int size, boolean allowConcatenatedGZIPStream)
+            throws IOException {
+        super(in, createInflater(in, size), size);
+        this.allowConcatenatedGZIPStream = allowConcatenatedGZIPStream;
+        usesDefaultInflater = true;
+        try {
+            readHeader(in);
+        } catch (IOException ioe) {
+            this.inf.end();
+            throw ioe;
+        }
     }
 
     /**
@@ -150,10 +166,45 @@ public int read(byte[] buf, int off, int len) throws IOException {
         }
         int n = super.read(buf, off, len);
         if (n == -1) {
-            if (readTrailer())
+            // reading of deflated data is now complete, we now read the GZIP stream trailer
+            readTrailer();
+            if (!allowConcatenatedGZIPStream) {
                 eos = true;
-            else
+            } else {
+                // This GZIPInputStream instance was created to allow potential
+                // concatenated GZIP stream, so we now try and read the next
+                // GZIP stream header, if any.
+
+                // use any un-inflated remaining bytes from the stream
+                InputStream headerIS = in;
+                int remainingUnInflated = inf.getRemaining();
+                if (remainingUnInflated > 0) {
+                    headerIS = new SequenceInputStream(
+                            new ByteArrayInputStream(this.buf, this.len - remainingUnInflated,
+                                    remainingUnInflated),
+                            new FilterInputStream(in) {
+                                public void close() throws IOException {}
+                            });
+                }
+                int numHeaderBytes;
+                try {
+                    numHeaderBytes = readHeader(headerIS); // next GZIP stream header, if any
+                } catch (EOFException _) {
+                    // TODO: somehow verify that this failed when reading the
+                    //  first byte of the header?
+                    eos = true;
+                    return n; // -1
+                }
+                // reset the inflater to read the deflated content of the next GZIP stream
+                inf.reset();
+                // adjust the input to the inflater correctly past the GZIP stream header
+                if (remainingUnInflated > numHeaderBytes) {
+                    inf.setInput(this.buf, this.len - remainingUnInflated + numHeaderBytes,
+                            remainingUnInflated - numHeaderBytes);
+                }
+                // read the next GZIP stream's deflated data
                 return this.read(buf, off, len);
+            }
         } else {
             crc.update(buf, off, n);
         }
@@ -242,12 +293,12 @@ private int readHeader(InputStream this_in) throws IOException {
      * reached, false if there are more (concatenated gzip
      * data set)
      */
-    private boolean readTrailer() throws IOException {
+    private void readTrailer() throws IOException {
         InputStream in = this.in;
-        int n = inf.getRemaining();
-        if (n > 0) {
+        int remainingUnInflated = inf.getRemaining();
+        if (remainingUnInflated > 0) {
             in = new SequenceInputStream(
-                        new ByteArrayInputStream(buf, len - n, n),
+                        new ByteArrayInputStream(buf, len - remainingUnInflated, remainingUnInflated),
                         new FilterInputStream(in) {
                             public void close() throws IOException {}
                         });
@@ -255,20 +306,15 @@ public void close() throws IOException {}
         // Uses left-to-right evaluation order
         if ((readUInt(in) != crc.getValue()) ||
             // rfc1952; ISIZE is the input size modulo 2^32
-            (readUInt(in) != (inf.getBytesWritten() & 0xffffffffL)))
+            (readUInt(in) != (inf.getBytesWritten() & 0xffffffffL))) {
             throw new ZipException("Corrupt GZIP trailer");
-
-        // try concatenated case
-        int m = 8;                  // this.trailer
-        try {
-            m += readHeader(in);    // next.header
-        } catch (IOException ze) {
-            return true;  // ignore any malformed, do nothing
         }
+        int numTrailerBytes = 8;
         inf.reset();
-        if (n > m)
-            inf.setInput(buf, len - n + m, n - m);
-        return false;
+        if (remainingUnInflated > numTrailerBytes) {
+            inf.setInput(buf, len - remainingUnInflated + numTrailerBytes,
+                    remainingUnInflated - numTrailerBytes);
+        }
     }
 
     /*
```
It still has one or two unanswered questions that I haven't fully thought of, but I thought this diff might be useful while discussing these changes.

Reply:
Hi @jaikiran,

>  I think the only new parameter we should introduce is whether or not the underlying input stream in is expected/allowed to have concatenated GZIP stream(s).

I disagree with you here. Based on what I think the goal is here, we need two separate flags.

Here's my thinking, please see if this makes sense to you...

The overall problem being addressed here is that this class does not give the user precise control of how the input is to be decoded. In particular, it fails in two ways:
* It doesn't give the user the ability to stop reading (precisely) after the first GZIP stream
* It doesn't give the user the ability to detect (some) erroneous inputs after the first GZIP stream

Detecting erronous input is just a basic capability that any "decoder" should have. For example, we would never accept a class file reader that failed to detect and report unknown bytecode.

While this class does correctly detect invalid GZIP data frames that appear at the start of the overall input, or in the middle of a compressed GZIP stream, in trying to be "smart" by automatically handling concatenated GZIP streams, it ends up failing to detect corruption where extra bytes are added to the end of the input.

Instead, it will randomly do one of the following:
  * Completely ignore the corruption, or
  * Throw an exception, or
  * Misinterpret the extra bytes as the valid start of a concatenated GZIP input stream, and then...
    * Return zero or more bytes that were never actually written, and then...
      * Completely ignore the corruption, or
      * Throw an exception

That behavior is totally bogus!

So what is the goal here?

The first goal is that we should exactly preserve the current behavior so as to not break anything out there.

The second goal is to allow the user to specify that they want to read and decode an _exact whole number_ of concatenated GZIP input streams (either 1 or N), or else get an exception - period.

Do you not agree?

If you agree on that goal, from what I can tell, it doesn't seem possible to achieve it, while also maintaining backward compatibility, without having two flags.

Your patch is an example - it may fail to detect and report trailing garbage when in `allowConcatenatedGZIPStream` mode  (noted by your comment "TODO: somehow verify that this failed when reading the first byte of the header?")